File: 2024/day10.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <cstdint>
#include <cstring>
#include <vector>
#include <map>
#include <set>
#include <queue>
#include <algorithm>
#include <numeric>

#include <utils.h>

using namespace std;

const string test(
"89010123\n"
"78121874\n"
"87430965\n"
"96549874\n"
"45678903\n"
"32019012\n"
"01329801\n"
"10456732");

bool DEBUG = false;
bool TEST = false;
#define DAY "day10"

typedef vector<string> StringVector;
typedef Point<short> point_t;

ShortMatrix parse( string src )
{
    ShortMatrix sb(1);
    for( auto c : src )
    {
        if( c == '\n' )
            sb.push_back( vector<short>() );
        else
            sb.back().push_back( c-'0');
    }
    return sb;
}

int WIDTH = -1;
int HEIGHT = -1;

point_t dirs[4] = { {-1,0}, {1,0}, {0,-1}, {0,1}};

struct Record {
    short x;
    short y;
    short c;
    Record( short _x=0, short _y=0, short _c=0 )
    : x(_x), y(_y), c(_c)
    {}
};

pair<int,int> part1( ShortMatrix & data )
{
    // Find the zeros.

    vector<point_t> zeros;
    for( short y = 0; y < HEIGHT; y++ )
        for( short x = 0; x < WIDTH; x++ )
            if( !data[y][x] )
                zeros.push_back( point_t({x,y}));
    
    queue<Record> queue;
    int part1 = 0;
    int part2 = 0;

    for( auto & z : zeros )
    {
        queue.emplace( Record(z.x,z.y,0) );
        set<point_t> solutions;
        while( !queue.empty() )
        {
            Record p = queue.front();
            queue.pop();

            short c = p.c+1;
            for( auto & d : dirs )
            {
                short x0 = p.x+d.x;
                short y0 = p.y+d.y;
                if( 0 <= x0 && x0 < WIDTH && 0 <= y0 && y0 < HEIGHT && data[y0][x0] == c )
                {
                    if( c == 9 )
                    {
                        part2 ++;
                        solutions.insert( point_t(x0,y0) );
                    }
                    else
                    {
                        queue.emplace( Record(x0, y0, c));
                    }
                }
            }
        }
        part1 += solutions.size();
    }
    return pair(part1, part2);
}


int main( int argc, char ** argv )
{
    while( *++argv )
    {
        string arg(*argv);
        if( arg == "debug")
            DEBUG = true;
        else if( arg =="test")
            TEST = true;
    }

    string input = TEST ? test : file_contents(DAY".txt");

    ShortMatrix data = parse(input);
    WIDTH = data[0].size();
    HEIGHT = data.size();

    auto [one, two] = part1( data );
    cout << "Part 1: " << one << "\n";
    cout << "Part 2: " << two << "\n";
}

File: 2024/day10.py
import os
import sys
from pprint import pprint
from collections import defaultdict
from itertools import permutations

test = """\
89010123
78121874
87430965
96549874
45678903
32019012
01329801
10456732"""

day = os.path.splitext(os.path.basename(__file__))[0]

TEST = 'test' in sys.argv
DEBUG = 'debug' in sys.argv

if TEST:
    data = test.splitlines()
else:
    data = open(day+'.txt').read().splitlines()

data = [[int(d) for d in row] for row in data]

WIDTH = len(data[0])
HEIGHT = len(data)

dirs = ( (-1,0), (1,0), (0,-1), (0,1) )

def part1(data):
    # Find the zeros.
    zeros = []
    for y,row in enumerate(data):
        for x,c in enumerate(row):
            if not c:
                zeros.append( (x,y) )
    
    queue = []
    part1 = 0
    part2 = 0
    for zx,zy in zeros:
        queue.append( (zx, zy, 0) )
        solutions = set()
        while queue:
            x,y,c = queue.pop(0)
            c += 1
            for dx,dy in dirs:
                x0,y0 = x+dx,y+dy
                if x0 in range(WIDTH) and y0 in range(HEIGHT) and data[y0][x0] == c:
                    if c == 9:
                        part2 += 1
                        solutions.add( (x0,y0) )
                    else:
                        queue.append( (x0, y0, c))
        part1 += len(solutions)
    return part1, part2

p1,p2 = part1(data)
print("Part 1:", p1)
print("Part 2:", p2)
